diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-ipq806x.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-ipq806x.c
index 6ae13dc19510..fc9209fa3d63 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-ipq806x.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-ipq806x.c
@@ -319,6 +319,16 @@ static int ipq806x_gmac_probe(struct platform_device *pdev)
 	/* Enable PTP clock */
 	regmap_read(gmac->nss_common, NSS_COMMON_CLK_GATE, &val);
 	val |= NSS_COMMON_CLK_GATE_PTP_EN(gmac->id);
+	switch (gmac->phy_mode) {
+		case PHY_INTERFACE_MODE_RGMII:
+			val |= NSS_COMMON_CLK_GATE_RGMII_RX_EN(gmac->id) |
+			       NSS_COMMON_CLK_GATE_RGMII_TX_EN(gmac->id);
+			break;
+		case PHY_INTERFACE_MODE_SGMII:
+			val |= NSS_COMMON_CLK_GATE_GMII_RX_EN(gmac->id) |
+			       NSS_COMMON_CLK_GATE_GMII_TX_EN(gmac->id);
+			break;
+	}
 	regmap_write(gmac->nss_common, NSS_COMMON_CLK_GATE, val);
 
 	if (gmac->phy_mode == PHY_INTERFACE_MODE_SGMII) {
